### PR TITLE
keyboard: fix missing release event when IME is enabled while key press 

### PR DESCRIPTION
**Pull request by tokyo4j**
After commit [e2189903](https://gitlab.freedesktop.org/wlroots/wlroots/-/commit/e21899037a97dcade40149f05ab54229c0018644) in wlroots, when ctrl-f is pressed in firefox with a IME client running, the following key-release event for "f" is not sent, thus "f" is repeated like "ffffffffff..." in the input box of firefox. This is because the key-release event for "f" is firstly relayed to the IME client and then sent via the virtual keyboard created by the IME client while the key-press event is sent via physical keyboard, and with [e2189903](https://gitlab.freedesktop.org/wlroots/wlroots/-/commit/e21899037a97dcade40149f05ab54229c0018644), key-release events without a corresponding key-press event on the same keyboard is not emitted to the compositor.

https://github.com/user-attachments/assets/b87e7e7f-c288-4f50-8902-361a6f0be678

So this PR fixes this problem by not relaying the key-release event to the IME client when the corresponding key-press event was previously sent to the application.

This follows sway's behavior:
https://github.com/swaywm/sway/blob/f293418d9d8a9e71fc52bc75a83d24ec2cc934c9/sway/input/keyboard.c#L548-L561

### CODE REVIEW

File: src/input/keyboard.c [hunk 1]
```diff
@@ -677,7 +686,8 @@ keyboard_key_notify(struct wl_listener *listener, void *data)
 				&& event->state == WL_KEYBOARD_KEY_STATE_PRESSED) {
 			start_keybind_repeat(seat->server, keyboard, event);
 		}
-	} else if (!input_method_keyboard_grab_forward_key(keyboard, event)) {
+	} else if (pressed_sent_key_is_released
```
Comment (johanmalm) [inline src/input/keyboard.c:689]: I find this logic hard to follow. Could we considering refactoring it with the aim of keeping `keyboard_key_notify()` easy on the brain. For example:

1. Could the `pressed_sent_key_is_released` logic go at the top of `input_method_keyboard_grab_forward_key()`.
2. Or maybe a `ime_grab()` wrapper containing both.

**Review by johanmalm — COMMENTED** [review]
Thanks. And great work on tracking this down!

Have added a suggestion to refactor.

Is it safe to use this when linking to a wlroots version which does not contain the fix?

**Comment by tokyo4j** [comment]
I've updated the patch with another approach that only touches `ime.c`, since my previous patch had an issue with normal applications like foot with IME enabled; when you press and release a character, the release event was not forwarded to the IME while the press event was forwarded, so the IME began sending repeated press events as it could observe the press event.

Instead, I added a `lab_set` that tracks the key-press events forwarded to IME. If a key-release event doesn't have corresponding key-press event in the `lab_set`, the key-release event is not forwarded. I also like how my new approach doesn't touch the normal code paths without IME.

> Is it safe to use this when linking to a wlroots version which does not contain the fix?

Yeah it should be safe. I tested with wlroots 0.18.1 and it works without any issue.

**Comment by johanmalm** [comment]
LGTM. Not tested.
